File: milo/ui/src/test_verdict/tools/utils/utils.test.ts
// Copyright 2023 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { parseTestResultName } from './utils';

describe('parseTestResultName', () => {
  it('should parse valid test result name', () => {
    const name = parseTestResultName(
      'invocations/inv%2fid/tests/test%2fid/results/result%2fid',
    );
    expect(name).toEqual({
      invocationId: 'inv%2fid',
      testId: 'test/id',
      resultId: 'result%2fid',
    });
  });

  it('should reject invalid test result name', () => {
    expect(() =>
      parseTestResultName(
        'invocatios/inv%2fid/tests/test%2fid/results/result%2fid',
      ),
    ).toThrow('invalid test result name');
  });
});

File: analysis/internal/changepoints/inputbuffer/utils_test.go
// Copyright 2024 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package inputbuffer

import "testing"

// Output as of June 2024 on Intel Skylake CPU @ 2.00GHz:
// BenchmarkMergeOrderedRuns-96    	  289780	      4080 ns/op	       2 B/op	       0 allocs/op
func BenchmarkMergeOrderedRuns(b *testing.B) {
	b.StopTimer()
	var hotBuffer, coldBuffer []Run

	// Consistently passing test. This represents ~99% of tests.
	for i := range 100 {
		hotBuffer = append(hotBuffer, Run{
			CommitPosition: int64(i),
			Expected:       ResultCounts{PassCount: 1},
		})
	}

	for i := range 2000 {
		coldBuffer = append(coldBuffer, Run{
			CommitPosition: int64(i),
			Expected:       ResultCounts{PassCount: 1},
		})
	}

	dest := make([]*Run, 0, 2100)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		MergeOrderedRuns(hotBuffer, coldBuffer, &dest)
	}
}
